x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushTopNToSource.java

```diff
@@ -7,55 +7,204 @@
-        PhysicalPlan plan = topNExec;
-        PhysicalPlan child = topNExec.child();
-        if (canPushSorts(child)
-            && canPushDownOrders(topNExec.order(), x -> LucenePushDownUtils.hasIdenticalDelegate(x, ctx.searchStats()))) {
+        Pushable pushable = evaluatePushable(topNExec, x -> LucenePushDownUtils.hasIdenticalDelegate(x, ctx.searchStats()));
+        return pushable.rewrite(topNExec);
+    }
+
+    /**
+     * Multiple scenarios for pushing down TopN to Lucene source. Each involve checking a combination of conditions and then
+     * performing an associated rewrite specific to that scenario. This interface should be extended by each scenario, and
+     * include the appropriate rewrite logic.
+     */
+    interface Pushable {
+        PhysicalPlan rewrite(TopNExec topNExec);
+    }
+
+    private static Pushable NO_OP = new NoOpPushable();
+
+    record NoOpPushable() implements Pushable {
+        public PhysicalPlan rewrite(TopNExec topNExec) {
+            return topNExec;
+        }
+    }
+
+    record PushableExchangeExec(ExchangeExec exchangeExec, EsQueryExec queryExec) implements Pushable {
+        public PhysicalPlan rewrite(TopNExec topNExec) {
+            var sorts = buildFieldSorts(topNExec.order());
+            var limit = topNExec.limit();
+            return exchangeExec.replaceChild(queryExec.withSorts(sorts).withLimit(limit));
+        }
+    }
+
+    record PushableQueryExec(EsQueryExec queryExec) implements Pushable {
+        public PhysicalPlan rewrite(TopNExec topNExec) {
             var sorts = buildFieldSorts(topNExec.order());
             var limit = topNExec.limit();
+            return queryExec.withSorts(sorts).withLimit(limit);
+        }
+    }
 
-            if (child instanceof ExchangeExec exchangeExec && exchangeExec.child() instanceof EsQueryExec queryExec) {
-                plan = exchangeExec.replaceChild(queryExec.withSorts(sorts).withLimit(limit));
-            } else {
-                plan = ((EsQueryExec) child).withSorts(sorts).withLimit(limit);
+    record PushableGeoDistance(FieldAttribute fieldAttribute, Order order, Point point) {
+        private EsQueryExec.Sort sort() {
+            return new EsQueryExec.GeoDistanceSort(fieldAttribute.exactAttribute(), order.direction(), point.getLat(), point.getLon());
+        }
+
+        private static PushableGeoDistance from(StDistance distance, Order order, Map<NameId, FieldAttribute> refAttributes) {
+            if (distance.left() instanceof Attribute attr && distance.right().foldable()) {
+                return from(attr, distance.right(), order, refAttributes);
+            } else if (distance.right() instanceof Attribute attr && distance.left().foldable()) {
+                return from(attr, distance.left(), order, refAttributes);
+            }
+            return null;
+        }
+
+        private static PushableGeoDistance from(
+            Attribute attr,
+            Expression literal,
```

Comment:
I could do that, but then would need to split the `makeGeometryFromLiteral` into two versions, one for pushdown and one for compute engine. Since Literal is foldable() and fold() just calls value(), this way seems cleaner and simpler. Also, I did not want to make this code sensitive to exactly when literal folding occurs, so we could move it around later, more easily.